### PR TITLE
Load balancing

### PR DESCRIPTION
**Pull request by demasterr**
Load balancing has been added (with in mind that task balancing is not yet implemented, so not all could be tested yet).

### CODE REVIEW

File: src/aws/instancemanager/instancemanager.py [hunk 1]
```diff
@@ -281,6 +295,25 @@ async def run(self):
...
+                    if overloaded:
+                        pass  # The workload should be balanced. IM should do nothing.
+                    else:
+                        # Kill an underloaded worker. There is not enough work.
+                        self._kill_instance(underloaded[0], instance_types=['worker'])
+                else:
+                    if overloaded:
+                        # We need a new worker. There is too much work.
+                        worker_id = self.start_worker()
+                        self.timewindow.add_empty(worker_id)
```
Comment (SanderWaij) [inline src/aws/instancemanager/instancemanager.py:312]: Will this new worker be started as well or does it do that automatically? (or is it already in here somewhere?)

File: src/aws/instancemanager/instancemanager.py [hunk 2]
```diff
@@ -281,6 +295,25 @@ async def run(self):
 
                 self.check_all_living()
 
+                # Check if some worker is underloaded or overloaded.
+                underloaded = self.timewindow.get_underloaded()
+                overloaded = self.timewindow.get_overloaded()
+                print("Overloaded: {}, Underloaded: {}".format(overloaded, underloaded))
+                if underloaded:
+                    if overloaded:
```
[the remark this answers is not in the document]
Comment (demasterr) [reply src/aws/instancemanager/instancemanager.py:303]: Both are lists. One worker may be overloaded, while another is underloaded at the same time.

File: src/aws/instancemanager/instancemanager.py (hunk 1 above)
Comment (demasterr) [reply src/aws/instancemanager/instancemanager.py:312]: `self.start_worker()` does the trick. The start signal to the worker itself comes as soon as the worker started.

File: src/aws/instancemanager/instancemanager.py [hunk 3]
```diff
@@ -355,9 +403,120 @@ def __init__(self, nodescheduler, host=con.HOST, port=con.PORT_IM):
...
+                    self._queue_overloaded(instance):
+                overloaded.append(instance)
+                jobs_running.append(self.queue_window)
+        if sort:
+            overloaded, _ = (list(t) for t in zip(*sorted(zip(overloaded, jobs_running))))
+        return overloaded
+
+    def get_underloaded(self) -> list:
+        """
+        Get a list of underloaded workers. Currently, workers without a job.
```
[the remark this answers is not in the document]
Comment (demasterr) [reply src/aws/instancemanager/instancemanager.py:489]: No currently it only detects workers with 0 jobs.

**Pull request merged**